lib/vm-agent.js

```diff
@@ -9,450 +9,696 @@
+                cb();
+                return;
+            }
 
-    Object.keys(sample).forEach(function (key) {
-        self.sample[key] = sample[key];
-    });
-};
+            assert.object(stash.vmobj, 'require VM object');
 
+            self.vmapiClient.updateVm(stash.vmobj, function _onVmUpdate(err) {
+                var doneUpdate = (new Date()).getTime();
 
-/*
- * On startup, sendFullSample updates all VMs on the server. This will be a
- * blocking call before we turn on the event listeners so we allow
- * vm-agent to report the full state of the server first. It also gives us time
- * for VMAPI to be fully ready and responding to requests before we keep going
- */
-VmAgent.prototype.sendFullSample = function (opts, callback) {
-    var self = this;
-    var log = this.log;
-    var vmapiClient = this.vmapiClient;
+                self.log.debug({
+                    action: 'VMAPI.updateVm',
+                    elapsed: (doneUpdate - startUpdate),
+                    err: err
+                }, 'completed VMAPI.updateVm() for updateVmapiVm()');
 
-    this.updateSample({}, function (err, sample) {
-        if (err) {
-            log.error(err, 'updateSample failed, cannot sendFullSample');
-            callback(err);
-            return;
+                cb(err);
+            });
+        }
+    ]}, function _updateVmComplete(err) {
+        if (!err || !self.ready) {
+            // on success we clear the retryDelay for next time.
+            if (self.retryDelays[vm_uuid]) {
+                clearTimeout(self.retryDelays[vm_uuid].timer);
+                delete self.retryDelays[vm_uuid];
+            }
 
-        } else if (Object.keys(sample).length === 0) {
-            log.warn('empty sample returned by vmadm lookup');
             callback();
             return;
         }
 
-        self.setSample(sample);
-
-        vmapiClient.updateServerVms(self.uuid, sample, function (vmapiErr) {
-            if (vmapiErr) {
-                log.error(vmapiErr, 'Could not sendFullSample');
-                return callback(vmapiErr);
-            }
-
-            return callback();
+        self.retryFailedUpdate(err, vm_uuid, function _retryFailedUpdateCb() {
+            callback(err);
         });
     });
 };
 
-
 /*
- * This function will call vmapiClient.updateVm(). When options.vm exists was
- * passed it means that a vm is tentatively destroyed and we double check that
- * there. In case the vm is not destroyed we reload its data with vmadm lookup.
- * At the end we call updateVm to update the vm data on VMAPI
+ * This is called when a VM has failed an update. It is responsible for waiting
+ * for (and incrementing) the delay and then re-queuing the update if there's
+ * not already a timer doing the same thing.
  */
-VmAgent.prototype.sendSample = function (options, callback) {
+VmAgent.prototype.retryFailedUpdate = // eslint-disable-line
+function retryFailedUpdate(err, vm_uuid, callback) {
```

Comment:
`retryFailedUpdate` doesn't really need to call back, does it? Maybe naming it `scheduleRetryUpdate` and not passing it a callback would better convey its semantics?